### PR TITLE
Verify Justified Block Hash

### LINKED ISSUE #721: Verify justified_block_hash is the hash of the block at the justified_slot
Beacon node only verifies justified_block is stored in the DB by doing a lookup using justified_block_hash. 
https://github.com/prysmaticlabs/prysm/blob/master/beacon-chain/types/block.go#L266-L271

Instead of verifying justified block has to be in the current chain, one more additional check needs to be implemented: Beacon node needs to verify the justified block's slot number is the same as justified_slot

From spec:
`Verify that justified_block_hash is the hash of the block in the current chain at the slot -- justified_slot`

### PR DESCRIPTION
**Pull request by JerryFireman**
TODO(you): choose "part of" or "resolves" and the associated github issue #.

Resolves #721 

---

# Description

TODO(you): Write why you are making the changes in this pull request.
Verify that the attestation's slot number is the same as the slot number of the block retrieved by the justified slot hash.

TODO(you): Write a summary of the changes you are making.
Added code to return error if the the attestation's slot number does not match the slot number of the block retrieved by the justified slot hash.

TODO(you): Link anything that would be helpful or relevant to the reviewers.

### CODE REVIEW

File: .idea/vcs.xml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (terenc3t) [inline .idea/vcs.xml]: Can we not check this in? Thanks

File: beacon-chain/types/block.go [hunk 2]
```diff
@@ -271,6 +272,19 @@ func (b *Block) isAttestationValid(attestationIndex int, db beaconDB, aState *Ac
 		return false
 	}
 
+	// Retrieve block from db using justified_block_hash and verify its slot number is the same as attestation justified slot number
```
Comment (terenc3t) [inline beacon-chain/types/block.go]: please add a period at the end of the comment

File: beacon-chain/types/block.go [hunk 3]
```diff
@@ -271,6 +272,19 @@ func (b *Block) isAttestationValid(attestationIndex int, db beaconDB, aState *Ac
...
+	rBlock, err := db.GetBlock(hash)
+	if err != nil {
+		log.Errorf("failed to get block: %v", err)
```
Comment (terenc3t) [inline beacon-chain/types/block.go]: shouldn't you also return false here?

File: beacon-chain/types/block.go [hunk 4]
```diff
@@ -271,6 +272,19 @@ func (b *Block) isAttestationValid(attestationIndex int, db beaconDB, aState *Ac
...
+	}
+	if attestation.JustifiedSlot != rBlock.SlotNumber() {
+		log.Errorf("the attestation's slot number is not the same as the slot number of the block retrieved by the justified slot hash. Attestation slot number: %v",
```
Comment (terenc3t) [inline beacon-chain/types/block.go]: Capitalize `the`

File: beacon-chain/types/block.go (hunk 3 above)
Comment (terenc3t) [inline beacon-chain/types/block.go]: Capitalize `failed`

**Comment by JerryFireman** [comment]
Hi Terence

I believe I have addressed all of your comments and updated the PR.

The tests passed but I noticed some errors in the types/block.go file in areas that I have not touched. I am assuming that someone else is making changes to the file. I hope it was not something that I did.

Please let me know how the updated PR looks and if there are any more issues for me to address.

Jerry

**Comment by terenc3t** [comment]
@JerryFireman I don't think those comments are addressed, can you double check?

File: .idea/misc.xml [hunk 5]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (terenc3t) [inline .idea/misc.xml]: I don't think we should be checking `.idea/` in

Comment (rauljordan) [reply .idea/misc.xml]: +1

File: beacon-chain/types/block.go [hunk 6]
```diff
@@ -271,6 +272,20 @@ func (b *Block) isAttestationValid(attestationIndex int, db beaconDB, aState *Ac
 		return false
 	}
 
+	// Retrieve block from db using justified_block_hash and verify its slot number is the same as attestation justified slot number.
```
Comment (rauljordan) [inline beacon-chain/types/block.go]: Line too long, please split for more readability

File: beacon-chain/types/block.go [hunk 7]
```diff
@@ -271,6 +272,20 @@ func (b *Block) isAttestationValid(attestationIndex int, db beaconDB, aState *Ac
...
+	rBlock, err := db.GetBlock(hash)
+	if err != nil {
+		log.Errorf("Failed to get block: %v", err)
+		return false
+	}
+	if attestation.JustifiedSlot != rBlock.SlotNumber() {
+		log.Errorf("The attestation's slot number is not the same as the slot number of the block retrieved by the justified slot hash. Attestation slot number: %v",
```
Comment (rauljordan) [inline beacon-chain/types/block.go]: see above comment ^

File: beacon-chain/types/block.go (hunk 6 above)
Comment (nisdas) [reply beacon-chain/types/block.go]: +1

File: beacon-chain/types/block.go [hunk 8]
```diff
@@ -271,6 +272,20 @@ func (b *Block) isAttestationValid(attestationIndex int, db beaconDB, aState *Ac
...
+			attestation.JustifiedSlot)
+		log.Errorf("Retrieved block slot number: %v",
```
Comment (nisdas) [inline beacon-chain/types/block.go]: This log is short enough to be put in one line, it doesnt have to be split into two.

Comment (nisdas) [reply beacon-chain/types/block.go]: Also the formatting for slot number should be `%d` not `%v`, the same applies in the log above

File: .idea/misc.xml (hunk 5 above)
Comment (nisdas) [reply .idea/misc.xml]: +1 , please remove the `.idea/` folder

Comment (prestonvanloon) [reply .idea/misc.xml]: Feel free to add `.idea` to `.gitignore` to preview this in the future

**Comment by prestonvanloon** [comment]
I've removed the `.idea` directory for you @JerryFireman :)

File: beacon-chain/types/block.go [hunk 9]
```diff
@@ -271,6 +272,21 @@ func (b *Block) isAttestationValid(attestationIndex int, db beaconDB, aState *Ac
...
+	rBlock, err := db.GetBlock(hash)
+	if err != nil {
+		log.Errorf("Failed to get block: %v", err)
+		return false
+	}
+	if attestation.JustifiedSlot != rBlock.SlotNumber() {
+		log.Errorf("The attestation's slot number is not the same as the slot number of the block retrieved by the justified slot hash. " +
+			"Attestation slot number: %d",
+			attestation.JustifiedSlot)
+		log.Errorf("Retrieved block slot number: %d",rBlock.SlotNumber())
```
Comment (prestonvanloon) [inline beacon-chain/types/block.go]: Please make this part of the error message on the previous line. This is one issue and only one error should be logged

**Comment by prestonvanloon** [comment]
I think you need to sign the CLA again. The bot dropped all of the data a little while ago. Sorry about that!

File: beacon-chain/types/block.go (hunk 9 above)
Comment (JerryFireman) [reply beacon-chain/types/block.go]: I think I resigned the CLA. If it did not work, please let me know and I will try again.

**Comment by JerryFireman** [comment]
By the way, do you see any value in adding more detailed instructions for working with git/github to your contributor's guidelines? I am far from an expert but I have been reading up and would be happy to take a stab at it if you would like.

**Comment by terenc3t** [comment]
> By the way, do you see any value in adding more detailed instructions for working with git/github to your contributor's guidelines? I am far from an expert but I have been reading up and would be happy to take a stab at it if you would like.

We would love that! Feel free to open up a PR. Thank you

**Comment by JerryFireman** [comment]
@terenc3t OK I will get working on it. By the way, should I open the PR now or wait until I have something ready for review?

**Comment by terenc3t** [comment]
> @terenc3t OK I will get working on it. By the way, should I open the PR now or wait until I have something ready for review?

When you have something for review. Thanks!

**Comment by JerryFireman** [comment]
OK will do. Thanks for clarifying.

**Comment by rauljordan** [comment]
Hi @JerryFireman do you have an update on this PR?

**Comment by JerryFireman** [comment]
Hi @rauljordan, I was thinking that where we left off Preston had one change and said he would make it hmself. But I am guessing he is busy so I will go ahead and do it. Will try to do it today, tomorrow at the latest.

**Comment by JerryFireman** [comment]
Hi @rauljordan @prestonvanloon I addressed Preston's latest comment by eliminating the redundant error and combining its message with the previous line. I believe I have now addressed all of the comments. Please let me know if I missed any or if you see any other areas that can be improved.

**Comment by JerryFireman** [comment]
@terenc3t, per our discussion I created a new pull request to update CONTRIBUTING.md with more detailed Git and Github workflow.

Also, @prestonvanloon did I correctly address your comment from 17 days ago on this PR with my change from 12 days ago? Please let me know if there is anything else for me to do.

**Comment by rauljordan** [comment]
Hey @JerryFireman, unfortunately a lot has changed in the spec related to this PR and we have been working on refactoring our code to align with latest changes. This functionality is out of sync now with latest spec requirements and would be best to close the PR for now. Thanks for your help on this and we'll do a better job at keeping you in the loop on which functions are least likely to change moving forward.

**Pull request closed without merging**